Distance estimation

OliverW10 opened the pull request:
Added a height based distance estimation based off this:
http://docs.limelightvision.io/en/latest/cs_estimating_distance.html
Sets some camera properties (exposure, auto white balance) from the code
The angle estimation is not final and there are some issues with the current annotated display.

Also includes images to test with all captured between 3 and 11 meters straight on to the power port

OliverW10 committed:
fixed power port finder bugs

OliverW10 committed:
fixed power port finder bugs

OliverW10 committed:
set properties via code

OliverW10 committed:
set properties via code

OliverW10 committed:
added sample images

OliverW10 committed:
added sample images

OliverW10 committed:
lowered magic number and inverted angle

OliverW10 committed:
lowered magic number and inverted angle

OliverW10 committed:
rebased off master and ran black

OliverW10 committed:
Merge branch 'distance_estimation' of https://github.com/OliverW10/vision-2020 into distance_estimation

OliverW10 committed:
changed test code to make testing easier

OliverW10 committed:
removed zooming argument in testing

auscompgeek commented on vision.py:
```diff
@@ -42,6 +42,13 @@ def __init__(
+        self.CameraManager.setCameraProperty(0, "exposure_auto", 1)
+        # self.CameraManager.setCameraProperty(0, "exposure_auto_priority", 0)
+        self.CameraManager.setCameraProperty(0, "focus_auto", 0)
+        self.CameraManager.setCameraProperty(0, "exposure_absolute", 1)
```
Doesn't the camera have a `setExposure` method?

auscompgeek commented on vision.py:
```diff
@@ -145,19 +152,33 @@ def create_annotated_display(
+        distance2 = self.translate(distance2, 45, 18, 3, 11)
+
+        combined_dist = (distance2 + distance) / 2
+        ajusted_dist = (
```
I think you typoed `adjusted`.

OliverW10 replied:
fixed

raulvera59 commented on vision.py:
```diff
@@ -145,19 +152,33 @@ def create_annotated_display(
+        target_angle = (
+            (y / FRAME_HEIGHT)
+        ) * MAX_FOV_HEIGHT  # vertical angle of the target from the center of the camera
+        distance = (2.5 - CAMERA_HEIGHT) / math.tan(GROUND_ANGLE + target_angle)
```
Shouldn't this 2.5 be TARGET_HEIGHT from magic_numbers.py?

OliverW10 committed:
temporarily removed eroding and dilating, fixed adjusted typo

raulvera59 commented on vision.py:
```diff
@@ -145,19 +152,33 @@ def create_annotated_display(
+        combined_dist = (distance2 + distance) / 2
+        ajusted_dist = (
+            5.21 + -1.15 * combined_dist + 0.145 * combined_dist ** 2
+        )  # gotten from google sheets
```
Again, what are these adjustments and why do we need them? This needs at least some comments and should use the numbers as constants with descriptive names.

raulvera59 reviewed (changes requested):
Unless I'm mistaken, the target height issue is a source of inaccuracy. But even if it isn't, the magic number should be replaced by a constant. The rest is just that I don't understand the new distance computation because it's too full of magic numbers. If it works, it works, but it would be better if it were clearer.

LucienMorey commented on vision.py:
```diff
@@ -145,19 +152,33 @@ def create_annotated_display(
+        distance = self.translate(distance, 1.7, 0.65, 3, 11)
+
+        distance2 = h
+        distance2 = self.translate(distance2, 45, 18, 3, 11)
```
What are distance and distance 2

LucienMorey commented on vision.py:
```diff
@@ -177,18 +198,20 @@ def get_image_values(self, frame: np.ndarray) -> tuple:
-        self.mask = cv2.erode(self.mask, None, dst=self.mask, iterations=1)
-        self.mask = cv2.dilate(self.mask, None, dst=self.mask, iterations=2)
-        self.mask = cv2.erode(self.mask, None, dst=self.mask, iterations=1)
+        # self.mask = cv2.erode(self.mask, None, dst=self.mask, iterations=1)
+        # self.mask = cv2.dilate(self.mask, None, dst=self.mask, iterations=2)
+        # self.mask = cv2.erode(self.mask, None, dst=self.mask, iterations=1)
```
Why are these commented out?

raulvera59 replied:
You shouldn't just resolve without answering the question here, even if you talked to Lucien about it. Otherwise the record here is incomplete.

LucienMorey commented on vision.py:
```diff
@@ -145,19 +152,33 @@ def create_annotated_display(
-        # print(f"Total Angle: {math.degrees(target_angle + GROUND_ANGLE)}")
-        return (TARGET_HEIGHT - CAMERA_HEIGHT) / math.tan(GROUND_ANGLE + target_angle)
+        return -(
+            ((X / FRAME_WIDTH) - 0.5) * MAX_FOV_WIDTH
```
What is the magic 0.5?

Injector-Spenral committed:
Corrected get_vertical_angle()

1st value for atan2 was wrong, I have updated it to correctly find the distance from the frame's vertical midpoint to `p`.

Injector-Spenral committed:
Updated get_horizontal_angle to use atan2 instead of linear approximation.

OliverW10 committed:
fixed camera and and changed defaults to [] rather than None

OliverW10 committed:
made POWRE_PORT_AREA_RATIO a magic number

OliverW10 committed:
bug in connection test

raulvera59 commented on connection.py:
```diff
@@ -11,15 +11,15 @@
             networktables entries (only if using_nt = True)
         """
-        if type(test) != type(None):
+        if test:
```
Just
self.test = test

raulvera59 commented on vision.py:
```diff
@@ -118,7 +120,7 @@ def find_power_port(self, frame: np.ndarray) -> tuple:
                 if (
                     area > MIN_CONTOUR_AREA
-                    and area / hull_area > 0.2
+                    and area / hull_area > POWRE_PORT_AREA_RATIO
```
POWER, not POWRE. This will crash as it is.

raulvera59 reviewed (approved):
You have one typo that will crash. I've put a comment there. Just fix that and test, then merge.

You should replace the angle and distance functions with Sam's utility ones once those are in the repo, but that can happen in another PR.

OliverW10 committed:
fixed powre to power typo
